[WIP] Stop using LIST nodes for PHI operand lists

mikedn opened the pull request.

BruceForstall commented:
@dotnet/jit-contrib

sandreenko commented on src/jit/gentree.h:
```diff
@@ -2199,6 +2198,92 @@ struct GenTree
+        return UseList(gtUses);
+    }
+
+#if DEBUGGABLE_GENTREE
```
What is the purpose of `DEBUGGABLE_GENTREE`?

mikedn replied:
Every `GenTree` class needs a default constructor to keep `GenTree::VtablePtr GenTree::GetVtableForOper(genTreeOps oper)` happy. It declares variables of various `GenTree` types so it can get its hands on a vtable pointer to use when the class of a `GenTree` node is changed using to `ChangeOper`. Of course, it's all a rather crazy, non-standard, not-anything hack :smile:

BruceForstall replied:
From a user perspective, it allows the VS debugger to show you the "real" subtype of a `GenTree*` variable, which is super useful. Maybe something that is done in a more standard way with RTTI?

mikedn replied:
I don't know, the standard way would be to use placement `new` to change the class of an object. But... I need to double check the standard... I think doing that will invalidate the existing object's contents and we don't want that when using `ChangeOper`.

Considering that this is debug only code I think it's fine as it is. Perhaps we should just add a macro to declare these default constructors on a single line, just to reduce debug noise.

BruceForstall replied:
One "problem" with `DEBUGGABLE_GENTREE` is it isn't available in Release, so not in production crash dumps, but making it available there was considered too expensive, and unnecessary.

mikedn replied:
> but making it available there was considered too expensive, and unnecessary

Yeah, at the end of day, `DEBUGGABLE_GENTREE` is pure convenience. One can still figure out the class from `gtOper`, it's just more work to do so in the debugger. I suppose it would be cool if C++'s RTTI was somehow customizable but it isn't, it requires a vtable pointer to work.

AndyAyersMS replied:
Maybe consider adding an operand count? There may be places we'd prefer to avoid traversing large PHI lists.

sandreenko commented on src/jit/gentree.h:
```diff
@@ -2239,8 +2324,11 @@ class GenTreeUseEdgeIterator final
-    int       m_state;
+    union {
+        GenTree*         m_argList;
+        GenTreePhi::Use* m_phiUseList;
```
From the first look I expected `m_phiUseList` to be `UseList`.

mikedn replied:
More like `UseIterator`. I went back and forth on this but I think it needs to be a pointer due `GenTreeUseEdgeIterator`'s `==` operator
```C++
        return (m_node == other.m_node) && (m_edge == other.m_edge) && (m_argList == other.m_argList) &&
               (m_state == other.m_state);
```
This won't work correctly unless the 2 union members have the same layout/meaning. And even then it's very close to, if not right into, the undefined behavior land.

mikedn replied:
I should add that in an ideal world `GenTreePhi::gtUses` should be `private`. But it seems that it requires quite a bit more work for not so many benefits, there are are only 3 remaining references to `gtUses` outside of `GenTreePhi`. At the end of the day I'm trying to make things reasonably better, not perfect.

CarolEidt replied:
This seems a bit dicey to me - at least it might be good to make this a named union, and at least somewhat abstract the comparison that is done by the `==` operator, so that it's clear that these must be "compatible" in this way.

AndyAyersMS replied:
Since you have the root node couldn't you check for the case where it's a PHI and handle it specially, say comparing the other union fields instead?

mikedn replied:
Yep, I need to do something about this, it's a bit odd.

mikedn replied:
> Since you have the root node couldn't you check for the case where it's a PHI and handle it specially, say comparing the other union fields instead?

That would work but I'm concerned that this will make `operator ==` even more complex than it is. Unfortunatelly, the C++ iterator model (and by extension range for) doesn't work so great when the state is complex. The repeated comparison with `end` can become costly.

I went for a more low tech approach - just use a `void*`. The few places that use this pointer can just cast to the right type.

sandreenko commented on src/jit/assertionprop.cpp:
```diff
@@ -2013,13 +2013,13 @@ AssertionIndex Compiler::optAssertionGenPhiDefn(GenTree* tree)
     }
 
-    GenTree* phi = tree->gtOp.gtOp2;
+    GenTree* phi = tree->gtGetOp2()->AsPhi();
```
`GenTreePhi* phi = `?

mikedn replied:
Hah, lame me, thanks!

mikedn commented:
@sandreenko Sorry, I didn't notice you already approved this.

I pushed a commit to cleanup that redundant `AsPhi`.

> The only concern is that it looks like it will be hard to reuse this code for other cases in #19876, won't it?

Some of it can be reused but it remains to be seen if it's not more trouble than it's worth. The `GT_FIELD_LIST` change looks very similar to this one, the difference is that its `Use` has some extra members:
```C++
struct Use
{
    GenTree*  op;
    Use*      next;
    unsigned  offset;
    var_types type;
};
```
So it has to be a different class but I hope that I can make both use classes inherit from a base `GenTreeUse` class that has `op` and `next`. And make `UseList` and `UseIterator` templates.

But for `GT_HWINTRINSIC` it's more like that I'll use an array of uses rather than a linked list. The reason is that most intrinsics have 1-3 operands and that can fit inside `GenTree`. Using a linked list would be overkill. Its use class does not need `next`, only `op`.

For calls I would expect using a linked list again, similar to `PHI` and `FIELD_LIST`. But then calls have 2 arg lists so things will be slightly different.

That said, the main goal is other form of reusal - such as `fgSetTreeSeqHelper` that can be rewritten to use `GenTreeVisitor` in like 10 lines of code. The less custom traversal code we have, the better. It's large and error prone.

CarolEidt commented on src/jit/gentree.h:
```diff
@@ -2199,6 +2198,92 @@ struct GenTree
+{
+    struct Use
+    {
+        GenTree* op;
```
I don't think we have a policy, guideline or even consensus around this, but I find node to be more meaningful than 'op', which at least in the JIT is sometimes used to mean an operator and sometimes an operand (and yes, we've got `op1` and `op2` all over the place, but still ...)

mikedn replied:
`node` sounds pretty good to me too, I'll update. Anyway, "use" and "operand" seem to overlap a bit.

CarolEidt commented on src/jit/gentree.h:
```diff
@@ -2199,6 +2198,92 @@ struct GenTree
     inline GenTree(genTreeOps oper, var_types type DEBUGARG(bool largeNode = false));
 };
 
+struct GenTreePhi : public GenTree
```
This needs a header with a general description of the node, as well as the constraints around how these are used, e.g.:
- Each of the `Use`s are `GT_LCL_VAR` nodes
- This node is always the rhs of an assignment, the lhs of which is also a `GT_LCL_VAR` node.

mikedn replied:
While I'm at it, I'll mention that I wanted to make `op` `GenTreePhiArg*` rather than `GenTree*`, to make it clear what node types are supported. Unfortunately there are many places that take the address of `op` and pass it around as `GenTree**` so this is impossible to do.

CarolEidt commented on src/jit/ssabuilder.cpp:
```diff
@@ -673,6 +673,99 @@ static GenTree* GetPhiNode(BasicBlock* block, unsigned lclNum)
+    // Make sure it isn't already present: we should only add each definition once.
+    for (GenTreePhi::Use& use : phi->Uses())
+    {
+        assert(use.op->AsPhiArg()->GetSsaNum() != ssaNum);
```
Perhaps also assert that it's not for the same `pred`?

mikedn replied:
>  Speaking of which, do you have data on how large these lists get in bad cases?

Here you go: https://1drv.ms/x/s!Av4baJYSo5pjgsQZfywTf5l3NUXitw 

Data from all FX assemblies. Largest PHI has 255 operands. 90% have only 2 operands.

I was actually considering handling the 2 operands case specially - store the 2 uses inside the `GenTreePhi` node itself. But the code required to handle this special case gets kind of ugly.

mikedn replied:
> It is more operands in some cases but perhaps avoids quadratic builder behavior for large phis.

I was curious to see what such quadratic behavior means on today hardware: https://gist.github.com/mikedn/23e708655cdb1d431ed165e300192c9e

It's less than 1μs for N = 10, ~68μs for N = 255, ~1ms for N = 1000 and ~200ms for N = 10000.

I'd say we're on the safe side.

mikedn replied:
> And if we just record the "first block" that brings in a value I wonder how RangeCheck::ComputeRange works correctly for PHIs as it seems like it would skip looking at some preds.

Indeed, that seems broken. If one predecessor asserts that `i <= 2` and another asserts that `i <= 3` the resulting PHI range should have upper bound 3. But if you miss the later predecessor then the upper bound will be only 2... Now, there may be something else in RangeCheck that prevents this from happening, though I starred a lot at RangeCheck in the past and I just don't see that something else.

mikedn replied:
> Perhaps also assert that it's not for the same pred?

Nope, it turns out that this assert isn't valid due to exception handling. It is possible to have something like `PHI BB05:V02:u2, BB05:V02:u3, BB05:V02:u4` when the same block in a try region contains more than one definition of the variable that's live in the handler.

It's the BB-SSA_NUM pair that's supposed to be unique in the PHI arg list. But since the JIT already requires SSA_NUM to be unique there's not much we can validate about the BB.

mikedn commented on src/jit/gentree.cpp:
```diff
@@ -1511,6 +1511,24 @@ bool GenTree::Compare(GenTree* op1, GenTree* op2, bool swapOK)
 
+        case GT_PHI:
+        {
+            GenTreePhi::UseIterator op1Use    = op1->AsPhi()->Uses().begin();
```
Might be useful to also mention that the order matters for equality, even though it's otherwise irrelevant. That's a bit peculiar but I don't think this code is ever used.

mikedn commented:
I think I addressed all review feedback with the exception of adding a use count to PHI. It's easy to do that, there's enough space in `GenTreePhi` and after this change the only point where PHI args are added is `SsaBuilder::AddPhiArg`.

But considering just how few PHIs have a large number of arguments adding the count seems premature. And there don't seem to be many places where we could skip large PHIs:
* `RangeCheck` - This could be useful but `RangeCheck` has its own chase budget that prevents from becoming too costly in extreme circumstances
* assertionprop's `optAssertionGenPhiDefn` could probably skip large PHIs, but it's not that costly
* VN might be an interesting case. The VN of a PHI is a chain of `VNF_Phi` VNs and 255 of them is pretty crazy. Perhaps it would make sense to fallback to a unique VN for large PHIs. But that should be investigated separately. And again, considering that so few PHIs are large...

RussKeldorph commented:
@dotnet/jit-contrib Take now or post-3.0?

AndyAyersMS commented:
I think we should hold off on taking an refactoring changes into 3.0.

At this point I think we're down to bug fixes and relatively contained and low-risk CQ improvements.

mikedn commented:
Also this is part of a series that will finish with the complete removal of `GT_LIST` so it's probably better and safer to have all or nothing in a release.

briansull commented:
> I think we should hold off on taking an refactoring changes into 3.0.

I also agree, we can take all  these changes post 3.0

mikedn commented:
The failed test output should be here: https://dev.azure.com/dnceng/public/_build/results?buildId=326410&view=ms.vss-test-web.build-test-results-tab&runId=9520886&resultId=100000&paneView=attachments

mikedn commented:
Best I can tell the test timed out after 10 minutes. History shows that it's run time is all over the place, from 1-2 minutes to 8-9 minutes. This kind of test to does not belong in Pri0 tests IMO...
